File: src/main/java/org/snipsniper/snipscope/ui/SnipScopeUIButton.java
package org.snipsniper.snipscope.ui;

import org.snipsniper.utils.IFunction;
import net.capturecoop.ccmathutils.Vector2Int;

import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class SnipScopeUIButton extends SnipScopeUIComponent{
    private final BufferedImage icon;
    private final BufferedImage iconPressed;
    private final BufferedImage iconHovering;

    private final ArrayList<IFunction> onPress = new ArrayList<>();
    private boolean isHovering = false;
    private boolean isHeld = false;
    private boolean selected = false;
    private Vector2Int lastPosition = new Vector2Int();

    public SnipScopeUIButton(BufferedImage icon, BufferedImage iconHovering, BufferedImage iconPressed) {
        this.icon = icon;
        this.iconHovering = iconHovering;
        this.iconPressed = iconPressed;
    }

    @Override
    public boolean mouseMoved(MouseEvent mouseEvent) {
        if(!super.mouseMoved(mouseEvent)) return false;
        isHovering = contains(mouseEvent.getPoint());
        lastPosition = new Vector2Int(mouseEvent.getPoint());
        return true;
    }

    @Override
    public boolean mouseDragged(MouseEvent mouseEvent) {
        if(!super.mouseDragged(mouseEvent)) return false;
        lastPosition = new Vector2Int(mouseEvent.getPoint());
        return true;
    }

    @Override
    public boolean mousePressed(MouseEvent mouseEvent) {
        if(!super.mousePressed(mouseEvent)) return false;
        if(isHovering)
            isHeld = true;
        return true;
    }

    @Override
    public boolean mouseReleased(MouseEvent mouseEvent) {
        if(!super.mouseReleased(mouseEvent)) return false;
        if(isHeld) {
            isHeld = false;
            if(contains(lastPosition.toPoint())) {
                for (IFunction function : onPress) {
                    function.run();
                }
            }
        }
        return true;
    }

    @Override
    public boolean render(Graphics2D g) {
        if(!super.render(g)) return false;
        BufferedImage toRender = icon;

        if(isHeld || selected)
            toRender = iconPressed;
        else if(isHovering)
            toRender = iconHovering;

        g.drawImage(toRender, getPosition().getX(), getPosition().getY(), getWidth(), getHeight(), null);
        return true;
    }

    public void setSelected(boolean bool) {
        selected = bool;
    }

    public void addOnPress(IFunction function) {
        onPress.add(function);
    }
}

File: src/main/kotlin/net/snipsniper/snipscope/ui/SnipScopeUIButton.kt
package net.snipsniper.snipscope.ui

import org.capturecoop.defaultdepot.math.Vector2I
import java.awt.Graphics2D
import java.awt.event.MouseEvent
import java.awt.image.BufferedImage

//TODO: This class is currently unused, but worked in the past, make sure it works if its used again
class SnipScopeUIButton(private val icon: BufferedImage, private val iconHovering: BufferedImage, private val iconPressed: BufferedImage) : SnipScopeUIComponent() {
    private val onPress = ArrayList<(() -> (Unit))>()
    private var isHovering = false
    private var isHeld = false
    private var selected = false
    private var lastPosition = Vector2I()

    override fun mouseMoved(mouseEvent: MouseEvent): Boolean {
        if (!super.mouseMoved(mouseEvent)) return false
        isHovering = contains(mouseEvent.point)
        lastPosition = Vector2I(mouseEvent.point)
        return true
    }

    override fun mouseDragged(mouseEvent: MouseEvent): Boolean {
        if (!super.mouseDragged(mouseEvent)) return false
        lastPosition = Vector2I(mouseEvent.point)
        return true
    }

    override fun mousePressed(mouseEvent: MouseEvent?): Boolean {
        if (!super.mousePressed(mouseEvent)) return false
        if (isHovering) isHeld = true
        return true
    }

    override fun mouseReleased(mouseEvent: MouseEvent?): Boolean {
        if (!super.mouseReleased(mouseEvent)) return false
        if (isHeld) {
            isHeld = false
            if (contains(lastPosition.toPoint())) {
                for (function in onPress) {
                    function.invoke()
                }
            }
        }
        return true
    }

    override fun render(g: Graphics2D): Boolean {
        if (!super.render(g)) return false
        var toRender = icon
        if (isHeld || selected) toRender = iconPressed else if (isHovering) toRender = iconHovering
        g.drawImage(toRender, position.x, position.y, width, height, null)
        return true
    }

    fun setSelected(bool: Boolean) {
        selected = bool
    }

    fun addOnPress(function: () -> (Unit)) {
        onPress.add(function)
    }
}